ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -83,2575 +83,1531 @@
-		if (p != NULL) {
-			while (n > 0) {
-				(*filter->filter_function)(*p++, filter);
-				n--;
-			}
+	size_t result = 0;
+	unsigned char *search_pos = haystack->val, *found_pos, *e = search_pos + haystack -> len;
+
+	while (search_pos < e) {
+		found_pos = (unsigned char*)zend_memnstr(
+			(const char*)search_pos,
+			(const char*)needle->val, needle->len,
+			(const char*)haystack->val + haystack->len);
+		if (!found_pos) {
+			break;
 		}
-		mbfl_convert_filter_delete(filter);
+		result++;
+		search_pos = found_pos + needle->len;
 	}
 
-	return len;
+	return result;
 }
 
-
 /*
- *  strpos
+ *  substr
  */
-struct collector_strpos_data {
-	mbfl_convert_filter *next_filter;
-	mbfl_wchar_device needle;
-	size_t needle_len;
-	size_t start;
-	size_t output;
-	size_t found_pos;
-	size_t needle_pos;
-	size_t matched_pos;
-};
+mbfl_string *mbfl_substr(mbfl_string *string, mbfl_string *result, size_t from, size_t length)
+{
+	const mbfl_encoding *encoding = string->encoding;
+
+	mbfl_string_init_set(result, encoding);
 
-static int
-collector_strpos(int c, void* data)
+	size_t start = char_offset_to_byte_offset(encoding, string->val, string->len, from);
+	size_t end   = char_offset_to_byte_offset(encoding, string->val + start, string->len - start, length) + start;
+
+	result->val = ecalloc(end - start, sizeof(unsigned char));
+	result->len = end - start;
+	memcpy(result->val, string->val + start, end - start);
+
+	return result;
+}
+
+/*
+ *  strcut
+ */
+static inline void strcut_const_width(mbfl_string *string, const mbfl_encoding *encoding, size_t from, const unsigned char **start, size_t *length)
 {
-	int *p, *h, *m;
-	ssize_t n;
-	struct collector_strpos_data *pc = (struct collector_strpos_data*)data;
-
-	if (pc->output >= pc->start) {
-		if (c == (int)pc->needle.buffer[pc->needle_pos]) {
-			if (pc->needle_pos == 0) {
-				pc->found_pos = pc->output;			/* found position */
-			}
-			pc->needle_pos++;						/* needle pointer */
-			if (pc->needle_pos >= pc->needle_len) {
-				pc->matched_pos = pc->found_pos;	/* matched position */
-				pc->needle_pos--;
-				goto retry;
-			}
-		} else if (pc->needle_pos != 0) {
-retry:
-			h = (int *)pc->needle.buffer;
-			h++;
-			for (;;) {
-				pc->found_pos++;
-				p = h;
-				m = (int *)pc->needle.buffer;
-				n = pc->needle_pos - 1;
-				while (n > 0 && *p == *m) {
-					n--;
-					p++;
-					m++;
-				}
-				if (n <= 0) {
-					if (*m != c) {
-						pc->needle_pos = 0;
-					}
-					break;
-				} else {
-					h++;
-					pc->needle_pos--;
-				}
-			}
-		}
+	/* If the requested `from` and `length` would give us a range of bytes which starts
+	 * or ends in the middle of a multi-byte character, push the start position and/or
+	 * the end position forwards so we get only complete characters. This means the extracted
+	 * range may include slightly more or less bytes than requested.
+	 *
+	 * It's easiest when each character is encoded with a constant number of bytes: */
+	int bytes_per_char;
```

Comment:
```suggestion
	size_t bytes_per_char;
```
To avoid having to remember what C's `size_t & int` promotion rules are...

Reply:
Good point, thanks.